crates/block-producer/src/tx_pool/mod.rs

```diff
@@ -0,0 +1,514 @@
+    ///
+    /// This means we track inflight data and in blocks completed after this block,
+    /// **excluding** this block.
+    stale_block: BlockNumber,
+
+    /// The number of the next block we hand out.
+    next_block: BlockNumber,
+
+    /// All transactions currently inflight.
+    ///
+    /// This includes those not yet processed, those in batches and blocks after the stale block.
+    tx_pool: BTreeMap<TransactionId, InflightTransaction>,
+
+    /// Transactions ready to be included in a batch.
+    ///
+    /// aka transactions whose parent transactions are already included in batches.
+    tx_roots: BTreeSet<TransactionId>,
+
+    /// The next batches ID.
+    next_batch_id: BatchId,
+
+    batch_pool: BTreeMap<BatchId, InflightBatch>,
+
+    /// Batches which are ready to be included in a block.
+    ///
+    /// aka batches who's parent batches are already included in blocks.
+    batch_roots: BTreeSet<BatchId>,
+
+    /// Blocks which are inflight or completed but not yet considered stale.
+    block_pool: BTreeMap<BlockNumber, Vec<BatchId>>,
+
+    /// The next block we expect to complete.
+    next_completed_block: BlockNumber,
+}
+
+impl TransactionPool {
+    /// Complete barring todos.
+    pub fn add_transaction(
+        mut self,
+        transaction: ProvenTransaction,
+        mut inputs: TransactionInputs,
+    ) -> Result<(), AddTransactionError> {
+        // Ensure inputs aren't stale.
+        if inputs.current_block_height <= self.stale_block.0 {
+            return Err(AddTransactionError::StaleInputs {
+                input_block: BlockNumber(inputs.current_block_height),
+                stale_limit: self.stale_block,
+            });
+        }
+
+        let account_update = transaction.account_update();
+
+        // Inflight transactions upon which this new transaction depends due to building on their
+        // outputs.
+        let mut parents = BTreeSet::new();
+
+        // Merge inflight state with inputs.
+        //
+        // This gives us the latest applicable state for this transaction.
+        // TODO: notes and nullifiers.
+        if let Some((state, parent)) = self.account_state.get(&account_update.account_id()) {
+            parents.insert(*parent);
+            inputs.account_hash = Some(*state);
+        }
+
+        // Verify transaction input state.
+        // TODO: update notes and nullifiers.
+        if inputs.account_hash.unwrap_or_default() != account_update.init_state_hash() {
+            return Err(AddTransactionError::InvalidAccountState {
+                current: inputs.account_hash.unwrap_or_default(),
+                expected: account_update.init_state_hash(),
+            });
+        }
+
+        // Transaction is valid, update inflight state.
+        // TODO: update notes and nullifiers.
+        let tx_id = transaction.id();
+        self.account_state
+            .insert(transaction.account_id(), (account_update.final_state_hash(), tx_id));
```

Comment:
If `account_states` was a bespoke struct, we could move this logic there as something like:
```rust
impl AccountStates {
    pub fn add_update(
        &mut self,
        update: &AccountUpdate,
        tx_id: TransactionId,
        store_state: Option<Digest>,
    ) -> Result<Vec<TransactionId>, AddTransactionError> {

    }
}
```